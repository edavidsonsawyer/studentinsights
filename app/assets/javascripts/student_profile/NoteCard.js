import React from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash';
import moment from 'moment';
import {linkBlue, strongOrange} from '../helpers/colors';
import {ERROR, PENDING} from '../helpers/requestStates';
import {formatEducatorName} from '../helpers/educatorName';
import Nbsp from '../components/Nbsp';
import NoteText from '../components/NoteText';
import NoteShell from '../components/NoteShell';
import Educator from '../components/Educator';
import NotifyAboutError from '../components/NotifyAboutError';
import EditableNoteText from '../components/EditableNoteText';
import RestrictedNotePresence from './RestrictedNotePresence';
import ModalSmall from './ModalSmall';


// This renders a single card for a Note of any type.
export default class NoteCard extends React.Component {
  constructor(props) {
    super(props);

    this.onTextChanged = this.onTextChanged.bind(this);
    this.debouncedSave = _.debounce(this.debouncedSave, 500);
  }

  // <EditableNoteText /> is uncontrolled, so it tracks
  // text changes, and syncs to save.
  debouncedSave(textValue) {
    if (!this.props.onSave) return;

    this.props.onSave({
      id: this.props.eventNoteId,
      eventNoteTypeId: this.props.eventNoteTypeId,
      text: textValue
    });
  }

  onTextChanged(textValue) {
    this.debouncedSave(textValue);
  }

  // No feedback, fire and forget
  onDeleteAttachmentClicked(eventNoteAttachmentId, e) {
    e.preventDefault();
    if (!confirm('Remove this link?')) return;
    this.props.onEventNoteAttachmentDeleted(eventNoteAttachmentId);
  }

  render() {
    const {noteMoment, badgeEl, educator, substanceOnly} = this.props;

    if (substanceOnly) return this.renderSubstanceEl();

    return (
      <NoteShell
        whenEl={noteMoment.format('MMMM D, YYYY')}
        badgeEl={badgeEl}
        educatorEl={<Educator educator={educator} />}
        substanceEl={this.renderSubstanceEl()}
      />
    );
  }

  renderSubstanceEl() {
    return (
      <div className="NoteCard-substance">
        {this.renderNoteSubstanceOrRedaction()}
        <div style={styles.footer}>
          {this.renderLastRevisedAt()}
          {this.renderRequestState()}
        </div>
        {this.renderAttachmentUrls()}
      </div>
    );
  }

  // For restricted notes, show a message and allow switching to another
  // component that allows viewing and editing.
  // Otherwise, show the substance of the note.
  //
  // If an onSave callback is provided, the text is editable.
  renderNoteSubstanceOrRedaction() {
    const {text, showRestrictedNoteRedaction, onSave} = this.props;

    if (showRestrictedNoteRedaction) {
      return this.renderRestrictedNoteRedaction();
    }
    if (!onSave) {
      return <NoteText text={text} />;
    }

    return (
      <EditableNoteText
        defaultText={text}
        onTextChanged={this.onTextChanged}
      />
    );
  }

  renderRestrictedNoteRedaction() {
    const {educator, urlForRestrictedNoteContent} = this.props;
    const educatorName = formatEducatorName(educator);
    const educatorFirstNameOrEmail = educatorName.indexOf(' ') !== -1
      ? educatorName.split(' ')[0]
      : educatorName;
    
    return (
      <RestrictedNotePresence
        studentFirstName={null} // The student name isn't passed
        educatorName={educatorFirstNameOrEmail}
        urlForRestrictedNoteContent={urlForRestrictedNoteContent}
      />
    );
  }

  renderLastRevisedAt() {
    const {nowFn} = this.context;
    const {lastRevisedAtMoment} = this.props;
    if (!lastRevisedAtMoment) return <Nbsp style={{fontSize: 12}} />;

    const now = nowFn();
    const revisedText = (now.clone().diff(lastRevisedAtMoment, 'days') < 45)
      ? lastRevisedAtMoment.from(now)
      : `on ${lastRevisedAtMoment.format('M/D/YY')}`;
    return (
      <span style={styles.revisionsText}>
        <span>Revised </span>
        {revisedText}
      </span>
    );
  }

  renderRequestState() {
    const {requestState} = this.props;

    if (requestState === ERROR) {
      return (
        <span style={styles.error}>
          <span>Your note is not saved</span>
          <ModalSmall
            icon={<span style={{color: linkBlue, fontWeight: 'normal'}}>Learn more</span>}
            style={{fontSize: 12}}
            modalStyle={{content: {}}}
            title="Your note is not saved"
            content={this.renderErrorSavingMessage()}
          />
        </span>
      );
    }

    if (requestState === PENDING) return <span style={styles.saving}>Saving...</span>;
    return <Nbsp style={{fontSize: 12}} />;
  }

  renderErrorSavingMessage() {
    return (
      <div>
        <div style={{color: strongOrange, fontWeight: 'bold'}}>There was an error communicating with the server to save your note.</div>
        <div style={{marginTop: 20, marginBottom: 10}} >To recover, try copying the latest version of the note, reloading the page, and pasting the changes.</div>
        <NotifyAboutError src="NoteCard#renderErrorSavingMessage" />
      </div>
    );
  }

  renderAttachmentUrls() {
    const {showRestrictedNoteRedaction, attachments} = this.props;
    if (showRestrictedNoteRedaction) return null;
    
    return attachments.map(attachment => {
      return (
        <div key={attachment.id}>
          <p style={{
            display: 'flex',
            alignItems: 'flex-start',
            marginTop: 10,
            fontSize: 12
          }}>
            <a
              href={attachment.url}
              target="_blank"
              rel="noopener noreferrer"
              style={{
                display: 'inline-block',
                marginRight: 10,
                fontSize: 12,
                overflow: 'hidden',
                textOverflow: 'ellipsis'
              }}>
              {attachment.url}
            </a>
            {this.renderRemoveAttachmentLink(attachment)}
          </p>
        </div>
      );
    });
  }

  // Can only remove attachments if callback is provided
  renderRemoveAttachmentLink(attachment) {
    if (!this.props.onEventNoteAttachmentDeleted) return null;

    return (
      <a
        href="#"
        onClick={this.onDeleteAttachmentClicked.bind(this, attachment.id)}
        style={{
          display: 'inline-block',
          marginLeft: 10,
          fontSize: 12,
          textDecoration: 'underline',
          color: '#999'
        }}>
        remove
      </a>
    );
  }
}
NoteCard.contextTypes = {
  nowFn: PropTypes.func.isRequired
};
NoteCard.propTypes = {
  noteMoment: PropTypes.instanceOf(moment).isRequired,
  badgeEl: PropTypes.node.isRequired,
  educator: PropTypes.object.isRequired,

  // Substance
  text: PropTypes.string.isRequired,
  lastRevisedAtMoment: PropTypes.instanceOf(moment),
  attachments: PropTypes.array.isRequired,

  // For editing eventNote only
  eventNoteId: PropTypes.number,
  eventNoteTypeId: PropTypes.number,
  onEventNoteAttachmentDeleted: PropTypes.func,
  onSave: PropTypes.func,
  requestState: PropTypes.string,

  // Configuring for different uses
  showRestrictedNoteRedaction: PropTypes.bool,
  urlForRestrictedNoteContent: PropTypes.string,
  substanceOnly: PropTypes.bool
};


const styles = {
  restrictedNoteRedaction: {
    color: '#999'
  },
  footer: {
    display: 'flex',
    justifyContent: 'space-between',
    alignItems: 'center',
    marginTop: 15
  },
  error: {
    fontSize: 12,
    color: strongOrange,
    fontWeight: 'bold',
    display: 'flex',
    alignItems: 'center'
  },
  saving: {
    fontSize: 12,
    color: '#aaa'
  },
  revisionsText: {
    color: '#aaa',
    fontSize: 12
  }
};
