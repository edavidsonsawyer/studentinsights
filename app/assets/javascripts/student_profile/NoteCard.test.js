import React from 'react';
import ReactDOM from 'react-dom';
import {studentProfile} from './fixtures/fixtures';
import {testTimeMoment} from '../testing/NowContainer';
import changeTextValue from '../testing/changeTextValue';
import NoteCard from './NoteCard';


const helpers = {
  renderInto(el, props) {
    const mergedProps = {
      noteMoment: testTimeMoment(),
      educatorId: 1,
      badge: <span>
        {''}
      </span>,
      onSave: jest.fn(),
      eventNoteId: 1,
      eventNoteTypeId: 1,
      educatorsIndex: studentProfile.educatorsIndex,
      attachments: [],
      ...props
    };

    return ReactDOM.render(<NoteCard {...mergedProps} />, el); //eslint-disable-line react/no-render-return-value
  },

  editNoteText(el, text) {
    const $text = $(el).find('.ResizingTextArea');
    changeTextValue($text.get(0), text);
  },

  getNoteHTML(el) {
    return $(el).find('.ResizingTextArea').html();
  }
};

describe('render', () => {
  it('renders simple text', () => {
    const el = document.createElement('div');
    helpers.renderInto(el, {
      text: 'hello'
    });

    expect(helpers.getNoteHTML(el)).toEqual('hello');
  });

  it('renders number of revisions', () => {
    const el = document.createElement('div');
    helpers.renderInto(el, {
      text: 'hello',
      numberOfRevisions: 1
    });

    expect($(el).text()).toContain('Revised 1 time');
  });

  it('escapes HTML-meaningful characters in text', () => {
    const el = document.createElement('div');
    helpers.renderInto(el, {
      text: 'hello <script src="xss.js"></script>world'
    });

    expect(helpers.getNoteHTML(el)).toEqual('hello &lt;script src="xss.js"&gt;&lt;/script&gt;world');
  });

  it('renders newlines', () => {
    const el = document.createElement('div');
    helpers.renderInto(el, {
      text: 'hello\nworld'
    });

    expect(helpers.getNoteHTML(el)).toEqual("hello\nworld");
  });
});

describe('saving', () => {
  it('works in simple case', () => {
    const el = document.createElement('div');
    const component = helpers.renderInto(el, { text: 'hello' });
    helpers.editNoteText(el, 'hello world');

    expect(component.props.onSave).toHaveBeenCalledWith({
      id: component.props.eventNoteId,
      eventNoteTypeId: component.props.eventNoteTypeId,
      text: 'hello world'
    });
  });
});
