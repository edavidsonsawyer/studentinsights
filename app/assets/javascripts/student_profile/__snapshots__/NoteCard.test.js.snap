// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
Array [
  <div>
    <div
      className="wrapper"
      style={
        Object {
          "display": "flex",
        }
      }
    >
      <div
        className="NoteCard"
        style={
          Object {
            "border": "1px solid #eee",
            "marginBottom": 10,
            "marginTop": 10,
            "padding": 15,
            "width": "100%",
          }
        }
      >
        <div
          style={
            Object {
              "display": "flex",
              "justifyContent": "space-between",
            }
          }
        >
          <span
            className="date"
            style={
              Object {
                "display": "inline-block",
                "fontWeight": "bold",
                "paddingRight": 10,
                "width": "11em",
              }
            }
          >
            February 22, 2018
          </span>
          <span />
          <span
            style={
              Object {
                "display": "inline-block",
                "paddingLeft": 5,
              }
            }
          >
            <a
              className="Educator"
              href="mailto:demo@example.com"
              style={Object {}}
            >
              demo@
            </a>
          </span>
        </div>
        <div
          style={
            Object {
              "marginTop": 10,
            }
          }
        >
          <div
            className="EditableNoteText"
          >
            <textarea
              className="ResizingTextArea"
              onChange={[Function]}
              style={
                Object {
                  "border": 0,
                  "borderRadius": 3,
                  "fontFamily": "'Open Sans', sans-serif",
                  "fontSize": 14,
                  "marginTop": 5,
                  "outline": "1px solid transparent",
                  "overflowX": "hidden",
                  "padding": 0,
                  "resize": "none",
                  "whiteSpace": "pre-wrap",
                  "width": "100%",
                  "wordBreak": "break-word",
                }
              }
              value="hello!"
            />
          </div>
          <div
            style={
              Object {
                "alignItems": "center",
                "display": "flex",
                "justifyContent": "space-between",
                "marginTop": 15,
              }
            }
          >
            <span>
               
            </span>
            <span>
               
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>,
  <div>
    <div
      className="wrapper"
      style={
        Object {
          "display": "flex",
        }
      }
    >
      <div
        className="NoteCard"
        style={
          Object {
            "border": "1px solid #eee",
            "marginBottom": 10,
            "marginTop": 10,
            "padding": 15,
            "width": "100%",
          }
        }
      >
        <div
          style={
            Object {
              "display": "flex",
              "justifyContent": "space-between",
            }
          }
        >
          <span
            className="date"
            style={
              Object {
                "display": "inline-block",
                "fontWeight": "bold",
                "paddingRight": 10,
                "width": "11em",
              }
            }
          >
            February 22, 2018
          </span>
          <span />
          <span
            style={
              Object {
                "display": "inline-block",
                "paddingLeft": 5,
              }
            }
          >
            <a
              className="Educator"
              href="mailto:demo@example.com"
              style={Object {}}
            >
              demo@
            </a>
          </span>
        </div>
        <div
          style={
            Object {
              "marginTop": 10,
            }
          }
        >
          <div
            className="EditableNoteText"
          >
            <textarea
              className="ResizingTextArea"
              onChange={[Function]}
              style={
                Object {
                  "border": 0,
                  "borderRadius": 3,
                  "fontFamily": "'Open Sans', sans-serif",
                  "fontSize": 14,
                  "marginTop": 5,
                  "outline": "1px solid transparent",
                  "overflowX": "hidden",
                  "padding": 0,
                  "resize": "none",
                  "whiteSpace": "pre-wrap",
                  "width": "100%",
                  "wordBreak": "break-word",
                }
              }
              value="hello!"
            />
          </div>
          <div
            style={
              Object {
                "alignItems": "center",
                "display": "flex",
                "justifyContent": "space-between",
                "marginTop": 15,
              }
            }
          >
            <span
              style={
                Object {
                  "color": "#aaa",
                  "fontSize": 12,
                }
              }
            >
              <span>
                Revised 
              </span>
              12 days ago
            </span>
            <span>
               
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>,
  <div>
    <div
      className="wrapper"
      style={
        Object {
          "display": "flex",
        }
      }
    >
      <div
        className="NoteCard"
        style={
          Object {
            "border": "1px solid #eee",
            "marginBottom": 10,
            "marginTop": 10,
            "padding": 15,
            "width": "100%",
          }
        }
      >
        <div
          style={
            Object {
              "display": "flex",
              "justifyContent": "space-between",
            }
          }
        >
          <span
            className="date"
            style={
              Object {
                "display": "inline-block",
                "fontWeight": "bold",
                "paddingRight": 10,
                "width": "11em",
              }
            }
          >
            February 22, 2018
          </span>
          <span />
          <span
            style={
              Object {
                "display": "inline-block",
                "paddingLeft": 5,
              }
            }
          >
            <a
              className="Educator"
              href="mailto:demo@example.com"
              style={Object {}}
            >
              demo@
            </a>
          </span>
        </div>
        <div
          style={
            Object {
              "marginTop": 10,
            }
          }
        >
          <div
            className="EditableNoteText"
          >
            <textarea
              className="ResizingTextArea"
              onChange={[Function]}
              style={
                Object {
                  "border": 0,
                  "borderRadius": 3,
                  "fontFamily": "'Open Sans', sans-serif",
                  "fontSize": 14,
                  "marginTop": 5,
                  "outline": "1px solid transparent",
                  "overflowX": "hidden",
                  "padding": 0,
                  "resize": "none",
                  "whiteSpace": "pre-wrap",
                  "width": "100%",
                  "wordBreak": "break-word",
                }
              }
              value="hello!"
            />
          </div>
          <div
            style={
              Object {
                "alignItems": "center",
                "display": "flex",
                "justifyContent": "space-between",
                "marginTop": 15,
              }
            }
          >
            <span
              style={
                Object {
                  "color": "#aaa",
                  "fontSize": 12,
                }
              }
            >
              <span>
                Revised 
              </span>
              on 12/19/16
            </span>
            <span>
               
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>,
  <div>
    <div
      className="wrapper"
      style={
        Object {
          "display": "flex",
        }
      }
    >
      <div
        className="NoteCard"
        style={
          Object {
            "border": "1px solid #eee",
            "marginBottom": 10,
            "marginTop": 10,
            "padding": 15,
            "width": "100%",
          }
        }
      >
        <div
          style={
            Object {
              "display": "flex",
              "justifyContent": "space-between",
            }
          }
        >
          <span
            className="date"
            style={
              Object {
                "display": "inline-block",
                "fontWeight": "bold",
                "paddingRight": 10,
                "width": "11em",
              }
            }
          >
            February 22, 2018
          </span>
          <span />
          <span
            style={
              Object {
                "display": "inline-block",
                "paddingLeft": 5,
              }
            }
          >
            <a
              className="Educator"
              href="mailto:demo@example.com"
              style={Object {}}
            >
              demo@
            </a>
          </span>
        </div>
        <div
          style={
            Object {
              "marginTop": 10,
            }
          }
        >
          <div
            className="EditableNoteText"
          >
            <textarea
              className="ResizingTextArea"
              onChange={[Function]}
              style={
                Object {
                  "border": 0,
                  "borderRadius": 3,
                  "fontFamily": "'Open Sans', sans-serif",
                  "fontSize": 14,
                  "marginTop": 5,
                  "outline": "1px solid transparent",
                  "overflowX": "hidden",
                  "padding": 0,
                  "resize": "none",
                  "whiteSpace": "pre-wrap",
                  "width": "100%",
                  "wordBreak": "break-word",
                }
              }
              value="hello!"
            />
          </div>
          <div
            style={
              Object {
                "alignItems": "center",
                "display": "flex",
                "justifyContent": "space-between",
                "marginTop": 15,
              }
            }
          >
            <span>
               
            </span>
            <span
              style={
                Object {
                  "color": "#aaa",
                  "fontSize": 12,
                }
              }
            >
              Saving...
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>,
  <div>
    <div
      className="wrapper"
      style={
        Object {
          "display": "flex",
        }
      }
    >
      <div
        className="NoteCard"
        style={
          Object {
            "border": "1px solid #eee",
            "marginBottom": 10,
            "marginTop": 10,
            "padding": 15,
            "width": "100%",
          }
        }
      >
        <div
          style={
            Object {
              "display": "flex",
              "justifyContent": "space-between",
            }
          }
        >
          <span
            className="date"
            style={
              Object {
                "display": "inline-block",
                "fontWeight": "bold",
                "paddingRight": 10,
                "width": "11em",
              }
            }
          >
            February 22, 2018
          </span>
          <span />
          <span
            style={
              Object {
                "display": "inline-block",
                "paddingLeft": 5,
              }
            }
          >
            <a
              className="Educator"
              href="mailto:demo@example.com"
              style={Object {}}
            >
              demo@
            </a>
          </span>
        </div>
        <div
          style={
            Object {
              "marginTop": 10,
            }
          }
        >
          <div
            className="EditableNoteText"
          >
            <textarea
              className="ResizingTextArea"
              onChange={[Function]}
              style={
                Object {
                  "border": 0,
                  "borderRadius": 3,
                  "fontFamily": "'Open Sans', sans-serif",
                  "fontSize": 14,
                  "marginTop": 5,
                  "outline": "1px solid transparent",
                  "overflowX": "hidden",
                  "padding": 0,
                  "resize": "none",
                  "whiteSpace": "pre-wrap",
                  "width": "100%",
                  "wordBreak": "break-word",
                }
              }
              value="hello!"
            />
          </div>
          <div
            style={
              Object {
                "alignItems": "center",
                "display": "flex",
                "justifyContent": "space-between",
                "marginTop": 15,
              }
            }
          >
            <span>
               
            </span>
            <span
              style={
                Object {
                  "alignItems": "center",
                  "color": "#ff9b00",
                  "display": "flex",
                  "fontSize": 12,
                  "fontWeight": "bold",
                }
              }
            >
              <span>
                Your note is not saved
              </span>
              <div
                className="click-event-modal"
                style={
                  Object {
                    "display": "inline",
                    "fontSize": 12,
                    "marginLeft": 10,
                  }
                }
              >
                <a
                  href="#"
                  onClick={[Function]}
                  style={
                    Object {
                      "fontSize": 12,
                      "outline": "none",
                    }
                  }
                >
                  <span
                    style={
                      Object {
                        "color": "#3177c9",
                        "fontWeight": "normal",
                      }
                    }
                  >
                    Learn more
                  </span>
                </a>
              </div>
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;
